src/web3-utils.js

```diff
@@ -147,10 +157,170 @@ export function formatUnits(
+        retryTimer = setTimeout(fetchTokenReserve, UNISWAP_MARKET_RETRY_EVERY)
+      }
+
+      return () => {
+        cancelled = true
+        clearTimeout(retryTimer)
+      }
+    }
+
+    fetchTokenReserve()
+  }, [tokenSymbol])
+
+  return useMemo(() => [tokenReserves, loading], [loading, tokenReserves])
+}
+
+// Wraps uniswap’s getMarketDetails() to only require
+// the other token in the pair, qualified using its symbol.
+async function getAnjMarketDetails(tokenSymbol) {
+  const [tokenData, anjData] = await Promise.all(
+    [tokenSymbol, 'ANJ'].map(symbol => {
+      // In the case of ETH, undefined should be passed as the reserves data.
+      if (symbol === 'ETH') {
+        return undefined
+      }
+
+      const [tokenAddress] = getKnownContract(`TOKEN_${symbol}`)
+      if (!tokenAddress) {
+        throw new Error(`Unsupported token symbol: ${symbol}`)
+      }
+      const tokenReserves = getTokenReserves(
+        tokenAddress,
+        Number(env('CHAIN_ID'))
+      )
+
+      if (!tokenReserves) {
+        throw new Error('Could not fetch reserves')
+      }
+      return tokenReserves
+    })
+  )
+
+  return getMarketDetails(tokenData, anjData)
+}
+
+// Wraps uniswap’s getTradeDetails() to only require
+// the other token in the pair, qualified using its symbol.
+async function getAnjTradeDetails(tokenSymbol, tradeAmount, fromAnj = false) {
+  const anjMarketDetails = await getAnjMarketDetails(tokenSymbol)
+  return getTradeDetails(
+    fromAnj ? TRADE_EXACT.OUTPUT : TRADE_EXACT.INPUT,
+    tradeAmount,
+    anjMarketDetails
+  )
+}
+
+// Fetch the rate at which a given token can get converted into ANJ.
+// Also returns the associated slippage.
+// Both are returned as BN.js instances.
+export function useAnjRate(symbol, amount, fromAnj = false) {
+  const [loading, setLoading] = useState(false)
+  const [rate, setRate] = useState(bigNum(0))
+  const [rateSlippage, setRateSlippage] = useState(bigNum(0))
+
+  useEffect(() => {
+    let cancelled = false
+    let retryTimer
+
+    if (amount === '0' || !amount) {
+      setLoading(false)
+      return
+    }
+
+    const updateRate = async () => {
+      try {
+        setLoading(true)
+
+        const {
+          executionRate,
+          executionRateSlippage,
+        } = await getAnjTradeDetails(symbol, amount, fromAnj)
+
+        const rate = fromAnj ? executionRate.rateInverted : executionRate.rate
+
+        if (!cancelled) {
+          setLoading(false)
+
+          setRate(
+            bigNum(
+              rate.isNaN() ? 0 : rate.times(10 ** UNISWAP_PRECISION).toFixed(0)
+            )
+          )
+
+          setRateSlippage(
+            bigNum(
+              executionRateSlippage.isNaN()
+                ? 0
+                : executionRateSlippage
+                    // convert from basis point to fractional
+                    .dividedBy(10000)
+                    .times(10 ** UNISWAP_PRECISION)
+                    .toFixed(0)
```

Comment:
A few questions:

1. If I remember correctly, the Uniswap SDK gives us back bignumber.js objects?
2. Does it ever return us NaNs?
3. What are the `10000` basis points?

Reply:
> If I remember correctly, the Uniswap SDK gives us back bignumber.js objects?

Yes, and the idea here is to convert them as early as possible into either a string or a BN.js instance.

> Does it ever return us NaNs?

I’m not sure but I don’t think it does?

> What are the `10000` basis points?

This is [the unit](https://en.wikipedia.org/wiki/Basis_point) Uniswap uses for `marketRateSlippage` and `executionRateSlippage`. See https://docs.uniswap.io/sdk-documentation/computation#example-usage-1

Reply:
> Yes, and the idea here is to convert them as early as possible into either a string or a BN.js instance.

Yes, agreed; it might be nice to leave a comment on this since the APIs are not compatible :(